pkg/streamingpromql/operators/binary_operation.go

```diff
@@ -665,22 +919,99 @@ func (b *binaryOperationSeriesBuffer) close() {
+type binaryOperationFunc func(lhs, rhs float64, hlhs, hrhs *histogram.FloatHistogram) (float64, *histogram.FloatHistogram, bool, error)
 
+// FIXME(jhesketh): Investigate avoiding copying histograms for binary ops.
 var arithmeticOperationFuncs = map[parser.ItemType]binaryOperationFunc{
-	parser.ADD: func(left, right float64) float64 {
-		return left + right
+	parser.ADD: func(lhs, rhs float64, hlhs, hrhs *histogram.FloatHistogram) (float64, *histogram.FloatHistogram, bool, error) {
+		if hlhs != nil && hrhs != nil {
+			res, err := hlhs.Copy().Add(hrhs)
+			if err != nil {
+				return 0, nil, false, err
+			}
+			return 0, res.Compact(0), true, nil
+		}
+		return lhs + rhs, nil, true, nil
+	},
+	parser.SUB: func(lhs, rhs float64, hlhs, hrhs *histogram.FloatHistogram) (float64, *histogram.FloatHistogram, bool, error) {
+		if hlhs != nil && hrhs != nil {
+			res, err := hlhs.Copy().Sub(hrhs)
+			if err != nil {
+				return 0, nil, false, err
+			}
+			return 0, res.Compact(0), true, nil
+		}
+		return lhs - rhs, nil, true, nil
+	},
+	parser.MUL: func(lhs, rhs float64, hlhs, hrhs *histogram.FloatHistogram) (float64, *histogram.FloatHistogram, bool, error) {
+		if hlhs != nil && hrhs == nil {
+			return 0, hlhs.Copy().Mul(rhs), true, nil
+		}
+		if hlhs == nil && hrhs != nil {
+			return 0, hrhs.Copy().Mul(lhs), true, nil
+		}
+		return lhs * rhs, nil, true, nil
+	},
+	parser.DIV: func(lhs, rhs float64, hlhs, hrhs *histogram.FloatHistogram) (float64, *histogram.FloatHistogram, bool, error) {
+		if hlhs != nil && hrhs == nil {
+			return 0, hlhs.Copy().Div(rhs), true, nil
+		}
+		return lhs / rhs, nil, true, nil
 	},
-	parser.SUB: func(left, right float64) float64 {
-		return left - right
+	parser.POW: func(lhs, rhs float64, hlhs, hrhs *histogram.FloatHistogram) (float64, *histogram.FloatHistogram, bool, error) {
+		return math.Pow(lhs, rhs), nil, true, nil
 	},
-	parser.MUL: func(left, right float64) float64 {
-		return left * right
+	parser.MOD: func(lhs, rhs float64, hlhs, hrhs *histogram.FloatHistogram) (float64, *histogram.FloatHistogram, bool, error) {
+		return math.Mod(lhs, rhs), nil, true, nil
 	},
-	parser.DIV: func(left, right float64) float64 {
-		return left / right
+	parser.ATAN2: func(lhs, rhs float64, hlhs, hrhs *histogram.FloatHistogram) (float64, *histogram.FloatHistogram, bool, error) {
+		return math.Atan2(lhs, rhs), nil, true, nil
 	},
-	parser.MOD:   math.Mod,
-	parser.POW:   math.Pow,
-	parser.ATAN2: math.Atan2,
+}
+
+// dataSeries couples data, sourceSeriesIndices, and sourceSeriesMetadata for use by mergeOneSide
+// Specifically it implements sort.Interface which will sort both data and sourceSeriesIndices at the same
+// time to retain a mapping of their indexes on either the present float timestamps or histograms as determined by sortByFloat.
+type dataSeries struct {
+	data                 []types.InstantVectorSeriesData
+	sourceSeriesIndices  []int
+	sourceSeriesMetadata []types.SeriesMetadata
```

Comment:
I think we should leave `sourceSeriesMetadata` out of this struct, and pass it around separately - `sourceSeriesMetadata` contains metadata for all series on one side of the operation, not just the series involved in this group, which is what the other two fields contain, so it might be a bit confusing.